packages/@lwc/ssr-runtime/src/styles.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { Stylesheets, Stylesheet } from '@lwc/engine-core';
+import { validateStyleTextContents } from './validate-style-text-contents';
+import type { LightningElementConstructor } from './lightning-element';
+
+function flattenStylesheets(stylesheets: Stylesheets): Array<Stylesheet> {
+    // recursively flatten
+    // @ts-expect-error we can optimize this function later
+    return stylesheets.filter(Boolean).flat(Infinity);
+}
```

Comment:
```suggestion
export function flattenStylesheets(stylesheets: Stylesheets): Stylesheet[] {
    const list: Stylesheet[] = [];
    for (const stylesheet of stylesheets) {
        if (!isArray(stylesheet)) {
            list.push(stylesheet);
        } else {
            list.push(...flattenStylesheets(stylesheet));
        }
    }
    return list;
}
```
Why not use the implementation from engine-core?

Reply:
```suggestion
export function flattenStylesheets(stylesheets: Stylesheets): Stylesheet[] {
    // Using a type assertion to avoid infinitely flattening an infinitely recursive type
    return (stylesheets as Stylesheet[]).flat(Infinity).filter(Boolean)
}
```
The type error is because `Stylesheets` is an infinitely nested array of arrays. Trying to flatten it to an infinite depth is what makes TypeScript explode. The type assertion avoids that.

Reply:
Moved `flattenStylesheets` from `@lwc/engine-core` into `@lwc/shared`.